active-rfcs/0000-style-script-behavior.md

```diff
@@ -0,0 +1,180 @@
+
+
+# Detailed design
+
+## `<style>`
+
+```astro
+<!-- INPUT: -->
+<style>h1 { color: red; }</style>
+
+<!-- OUTPUT: -->
+<style>h1.astro-ABCDEF { color: red; }</style>
+```
+
+### Current Behavior
+
+- Style content is scoped to the component.
+  - opt-out of this with `<style global>`, see below
+- Style content is processed, ex: Sass is supported.
+- Style content is added to the module graph and bundled with the rest of the page CSS.
+- `@import` will be resolved using Vite's current `@import` resolve logic.
+- No `<style>` tag is output in the final component HTML.
+- No HTML attributes supported on the element, since the element is virtual and never output.
+
+### New RFC Behavior
+
+- Cannot be nested within a template expression or conditional. Bundled styles must be scanned by the compiler. This is currently easy to break in v0.21 with something like `{alwaysFalse && <style>...` so this RFC moves to remove support for this. This change shouldn't impact many users.
+- Can only exist top-level in the template (best for components) or nested directly inside of `<head>` (best for pages/layouts). This is to guarentee that scanning and reading hoisted script content is straightforward and bug-free. This shouldn't affect many users.
+
+### Deep Dive: define:vars
+
+Even though `<style define:vars={...}>` is all about adding dynamic content, it should still continue to work with statically bundled CSS. This is because the CSS variable is referenced in the static `<style>` contents via `var(--foo)`, which can be safely bundled with the rest of your page CSS. Then, during the page render at runtime, the component will add a `<style>` tag to the page that defines the `--foo` variable.
+
+Note that this will continue to have issues with some use-cases. For example, if a component is used twice on the page, the dynamic value may change across different renders but would impact all components on the page. You can see an example of this here, where the last rendered component wins: https://stackblitz.com/edit/github-eww5sz?file=src%2Fcomponents%2FTour.astro&on=stackblitz
+
+
+## `<style global>`
+
+```astro
+<!-- INPUT: -->
+<style>h1 { color: red; }</style>
+
+<!-- OUTPUT: -->
+<style>h1 { color: red; }</style>
+```
+
+### Current Behavior
+
+- Same behavior described above, but not scoped to the component.
+
+## `<script>`
+
+```astro
+<!-- INPUT: -->
+<script>console.log('fired!');</style>
+
+<!-- OUTPUT: -->
+<script>console.log('fired!');</style>
+```
+
+### Current Behavior
+
+- Style content is escaped as JS but otherwise untouched by Astro. ex: TypeScript not supported.
+- ESM import will not be resolved or bundled, sent untouched/raw to the browser.
+- Will be output in the final component render(), useful for adding global JS.
+- May be duplicated if this component appears on the page multiple times.
+  - In the future, we could add some kind of of `is:unique` support to prevent duplicate scripts.
```

Comment:
Isn't this provided by `hoist`?

Reply:
yes, good point. This is trying to say that we could do `is:unique` seperately from hoist. But I guess none of that is really in scope for this RFC anymore, so maybe best to remove entirely.